idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,230 @@
+            - dsdp_dic(dict)                 : Gradient vector of the (decision variables, parameters) with respect to paramerters (=theta_name). e.g) dict = {'d(x1)/d(p1)': 1.0, 'd(x2)/d(p1)': 0.0, 'd(p1)/d(p1)': 1.0, 'd(p2)/d(p1)': 0.0, 'd(x1)/d(p2)': 0.0, 'd(x2)/d(p2)': 1.0, 'd(p1)/d(p2)': 0.0, 'd(p2)/d(p2)': 1.0}
+            - propagation_f(dictionary)      : Error propagation in the objective function with the dictionary key, 'objective', df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+            - propagation_c(dictionary)      : Error propagation in the constraints with the dictionary keys, 'constraints r' where r is the line number. If no constraint includes uncertain parameters, return an empty dictionary      
+
+    Raises:
+        Exception:  When model_uncertain is neither 'ConcreteModel' nor 'function'.
+    """
+
+    # define Pyomo model
+    if type(model_uncertain).__name__ == 'ConcreteModel':
+        model = model_uncertain
+    elif type(model_uncertain).__name__ == 'function':
+        model = model_uncertain()
+    else:
+        raise TypeError('model_uncertain must be either python function or Pyomo ConcreteModel.')
+    # Remove all "'" in theta_names     
+    theta_names, var_dic,variable_clean = clean_variable_name(theta_names)
+    for v in theta_names:
+        eval('model.'+var_dic[v]).setlb(theta[v])
+        eval('model.'+var_dic[v]).setub(theta[v])
+    # get gradient of the objective function, constraints, and the column number of each theta
+    gradient_f,gradient_f_dic, gradient_c,gradient_c_dic, line_dic = get_dfds_dcds(model, theta_names, tee)
+    dsdp_dic, col  = get_dsdp(model, theta_names, theta, var_dic,tee)      
+    x_list = [x for x in col if x not in var_dic.keys()]
+    p_list = theta_names #This makes cov and fxxp have the same order 
+    fxxp = []
+    for i in p_list:
+        fxxp_tmp = 0
+        for j in x_list:
+            fxxp_tmp = fxxp_tmp + gradient_f_dic['d(f)/d('+j+')']*dsdp_dic['d('+j+')/d('+i+')']
+        fxxp.append(fxxp_tmp)
+    fxxp = np.array(fxxp)
+    # calculate error propagation df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+    # objective function: 
+    for cc in list(model.component_data_objects(Objective,
+                                            active=True,
+                                            descend_into=True)):
+        # save gradient of the objective with respect to only theta_names
+        if any(ele in str(cc.expr) for ele in theta_names):
+            gradient_f_theta = gradient_f[[int(line_dic[v]) -1 for v in theta_names]]
+    propagation_f = {}
+    # calculate error propagation of the objective fuction
+    if 'gradient_f_theta' in locals():
+        propagation_f['objective'] = np.dot(gradient_f_theta,np.dot(cov,np.transpose(gradient_f_theta))) + np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    else:
+        propagation_f['objective'] = np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
```

Comment:
Why is propagation_f a dictionary? Why not just make it a scalar?

Reply:
Here, no clear reason. We can change it to a float.

Reply:
Changed dictionary to numpy.float64.